### ISSUE TITLE
Channels are 12x slower on Android versus iOS

### ISSUE THREAD

**Issue by gaaclarke** [flutter/flutter#53311]
Microbenchmarks from https://github.com/gaaclarke/MessagingTiming

Things are going to be a bit slower because of JNI, 12x seems excessive.

**Android:**
![Screenshot_20200325-173045](https://user-images.githubusercontent.com/30870216/77591755-e0e93980-6ead-11ea-8bc8-6188f5a31b5d.jpg)

**iOS:**
![IMG_4399](https://user-images.githubusercontent.com/30870216/77591797-f8c0bd80-6ead-11ea-80f4-a1a5bea5c9d8.PNG)

**Comment by xster** [comment]
cc @mit-mit

**Comment by mit-mit** [comment]
not sure who owns this these days; @csells ?

**Comment by gaaclarke** [comment]
I profiled a bit and chatted with @chinmaygarde.  It seems that a lot of time is spent in timerfd which is necessary on android in order to schedule with the android main event loop (platform thread).  So, it sounds like having a solution that doesn't synchronize with the main event loop may be our only option there.  Maybe we just lean into FFI.

I'm not great at android profiling, there may be some other minor improvements we can do if I can tease out a meaningful trace.

<img width="759" alt="Screen Shot 2020-03-25 at 4 13 42 PM" src="https://user-images.githubusercontent.com/30870216/77668896-40445980-6f41-11ea-9809-8fbf286b0c61.png">

**Comment by jason-simmons** [comment]
Something unusual is happening related to scheduling of the platform thread.

If I run this benchmark on a Pixel 3, I get these results (in microseconds):
```
  Simple (1st run): 114
  Simple (2nd run): 484
  Basic (1st run): 1001
  Basic (2nd run) 1119
```

The round trip time for a platform channel message starts off relatively fast but gets slower as the app runs.

I tried instumenting `Shell::OnEngineHandlePlatformMessage` and found that the delay between posting the task to the platform thread and executing the task can vary widely.  Initially the delay is a few tens of microseconds, but during the slower phases of the benchmark it rises to several hundred microseconds.

If I drag a finger on the screen while the benchmark executes, then the message round trip time is fast and and is consistent across all phases:
```
  Simple (1st run): 112
  Simple (2nd run): 107
  Basic (1st run): 106
  Basic (2nd run) 105
```

It seems like the tasks queued to the platform thread are being executed more promptly if the platform thread is receiving UI events.  Perhaps there is some signal that the engine could send to Android indicating that we want more aggressive scheduling of the platform thread?

**Comment by xster** [comment]
> It seems that a lot of time is spent in timerfd which is necessary on android in order to schedule with the android main event loop

What does that mean specifically? Just that timerfd is acting as a condition variable waiting for the Android main thread right?

I can't try the example. Is native_add.cpp referenced in https://github.com/gaaclarke/MessagingTiming/blob/master/android/CMakeLists.txt#L9 checked in? 

The variable timing observation is super interesting. I think comparing systraces between the 2 cases could be revealing since we can see what the CPU is doing for which thread at which priority during the whole process to see if it's from our [configurations of the threads](https://cs.opensource.google/flutter/engine/+/master:shell/platform/android/android_shell_holder.cc;l=137) or just a CPU governor issue

**Comment by xster** [comment]
I think a `adb shell ps -T -p [pid]` could be interesting to see whether our priority setting is working. Also I assume we don't really know if the slowness comes from dart->platform main or platform main->dart right? Since platform main has a higher priority than we can bring our thread to without rooting, splitting the 2 directions could be interesting (perhaps letting the platform side just save the timestamps with a key somewhere and summarizing later when it's not time critical). 

Also a quick `adb shell cat /sys/devices/system/cpu/cpu0/cpufreq/scaling_cur_freq` does show that touching and not touching the screen gets a different CPU frequency. I didn't test all the CPUs but could be a factor. If you're super ambitious, we can programmatically save the current frequency during the benchmark too to see the correlation.

**Comment by gaaclarke** [comment]
@xster Android might be broken for MessageTiming right now because I'm in the middle of implementing ffi, going back to 9c443a62a70f0e7c20452a156f248c8c30ebf58f should work.  I've taken a few day detour to do roll duties.

**Comment by gaaclarke** [comment]
@jason-simmons I was poking around at the android API and I noticed this: https://developer.android.com/reference/android/os/Handler#createAsync(android.os.Looper)

`Create a new Handler whose posted messages and runnables are not subject to synchronization barriers such as display vsync.`

If we were using the equivalent to`new Handler` versus `Handler.createAsync` that could create artificial delays in their processing.  I'm not familiar with what we are using for channels currently.

**Comment by jason-simmons** [comment]
Tasks queued to the engine's platform thread task runner are dispatched through a TimerFD added to the looper at the NDK level.
(see https://github.com/flutter/engine/blob/master/fml/platform/android/message_loop_android.cc#L47)

The task for an incoming platform message will then call the `FlutterJNI.handlePlatformMessage` hook via JNI.  Processing the message should not involve an `android.os.Handler` at the Java level.

**Comment by xster** [comment]
oh ha, that's what Aaron and I were looking for in terms of creating that behavior in his test :)

Turns out that's what we're doing in our engine too

**Comment by gaaclarke** [comment]
@jason-simmons Right, but the fact there is a distinction in the Java API points to maybe there is an equivalent in the NDK.  I'll read through the docs to see if i notice anything.

**Comment by gaaclarke** [comment]
@filiph has a nice writeup on scale locking to address the problem with inconsistent android performance benchmarks in his article:
https://medium.com/flutter/performance-testing-of-flutter-apps-df7669bb7df7

**Comment by chinmaygarde** [comment]
Triage: I don't see this tied to a customer blocking issue nor is it a new feature. Speculatively assigning P6 but feel free to revise.

**Comment by gaaclarke** [comment]
@chinmaygarde meta comment: this investigation was done as the result of feedback to @xster and @zoeyfan.  We need to come up with a better way to link and prioritize work to the results of their conversations and the feedback provided to the UXR team.
